src/NHibernate/AdoNet/ConnectionManager.cs

```diff
@@ -426,14 +426,17 @@ public DbCommand CreateCommand()
 		/// <param name="transaction">The transaction in which the connection should be enlisted.</param>
 		public void EnlistIfRequired(System.Transactions.Transaction transaction)
 		{
-			if (_connection == null ||
-				_connectionReleaseRequired ||
-				!ShouldAutoJoinTransaction ||
-				// Most connections do not support enlisting in a system transaction while already participating
-				// in a local transaction. They are not supposed to be mixed anyway.
-				IsInActiveExplicitTransaction ||
-				// This is not a reference equal, == operator is overridden for that type. Thus testing last.
-				transaction == _connectionAmbientTransaction)
+			if (transaction == _currentSystemTransaction)
+				return;
+
+			_currentSystemTransaction = transaction;
+
+			// Most connections do not support enlisting in a system transaction while already participating
+			// in a local transaction. They are not supposed to be mixed anyway.
+			if (IsInActiveExplicitTransaction && transaction != null)
+				throw new InvalidOperationException("Cannot enlist in a system transaction while an explicit transaction has been started on the session.");
```

Comment:
I think it is better to fail explicitly in this case, as would do `Connection.EnlistTransaction`.